File: src/detector.py
from pathlib import Path

import cv2
import numpy as np
import onnxruntime as ort


class YOLODetector:

    def __init__(self, model_path, inference_size=640, threads=1):
        """
        Load the YOLO ONNX model.

        Args:
            model_path:
                Path to the ONNX model file.

            inference_size:
                Square input size the model was exported at
                (e.g. 640 or 320). Used to resize the frame
                and map detections back to the original size.

            threads:
                ONNX runtime CPU threads (e.g. 4 on the
                Pi Zero 2 W). A big speedup for convolutions.
        """

        self.model_path = Path(model_path)

        self.inference_size = inference_size

        if not self.model_path.exists():
            raise FileNotFoundError(
                f"Model file not found: {self.model_path}"
            )

        # Load ONNX model with the CPU thread pool tuned for
        # the board. The default single thread heavily wastes
        # a quad-core Pi.

        session_options = ort.SessionOptions()

        session_options.intra_op_num_threads = int(threads)

        session_options.graph_optimization_level = (
            ort.GraphOptimizationLevel.ORT_ENABLE_ALL
        )

        self.session = ort.InferenceSession(
            str(self.model_path),
            sess_options=session_options,
            providers=["CPUExecutionProvider"]
        )

        print("YOLO model loaded successfully")


        # Get input information
        model_input = self.session.get_inputs()[0]

        self.input_name = model_input.name
        self.input_shape = model_input.shape

        print("Input name:", self.input_name)
        print("Input shape:", self.input_shape)


        # Get output information
        outputs = self.session.get_outputs()

        print("Number of outputs:", len(outputs))

        for output in outputs:
            print(
                "Output:",
                output.name,
                output.shape
            )


    def predict(self, frame):
        """
        Run YOLO inference on one frame.
        """

        # Resize frame to YOLO input size
        resized_frame = cv2.resize(
            frame,
            (self.inference_size, self.inference_size)
        )


        # Convert BGR to RGB
        rgb_frame = cv2.cvtColor(
            resized_frame,
            cv2.COLOR_BGR2RGB
        )


        # Normalize pixels
        normalized_frame = (
            rgb_frame.astype(np.float32) / 255.0
        )


        # HWC -> CHW
        input_tensor = np.transpose(
            normalized_frame,
            (2, 0, 1)
        )


        # Add batch dimension
        # (3,640,640) -> (1,3,640,640)

        input_tensor = np.expand_dims(
            input_tensor,
            axis=0
        )


        # Run inference

        outputs = self.session.run(
            None,
            {
                self.input_name: input_tensor
            }
        )


        return outputs[0]



    def detect_person(self, frame, confidence_threshold=0.5):
        """
        Detect the person with highest confidence.

        Returns:
            Dictionary with:
            - confidence
            - bounding box coordinates

            or None
        """


        # Run YOLO

        output = self.predict(frame)


        # Remove batch dimension
        # (1,84,8400)
        #      |
        #      v
        # (84,8400)

        predictions = output[0]


        # Change:
        # (84,8400)
        #
        # to:
        #
        # (2100,84)

        predictions = predictions.T


        person_class_id = 0


        # Person confidences for every anchor (vectorised,
        # much faster than a Python loop).

        confidences = predictions[:, 4 + person_class_id]

        best_index = int(np.argmax(confidences))

        best_confidence = float(confidences[best_index])


        if best_confidence < confidence_threshold:

            return None


        # Bounding box in YOLO format

        x_center, y_center, width, height = predictions[best_index, :4]


        # Convert center/width/height to corners, then scale
        # from model coordinates back to the original frame.

        original_height, original_width = frame.shape[:2]

        scale_x = original_width / self.inference_size
        scale_y = original_height / self.inference_size

        x1 = int((x_center - width / 2) * scale_x)
        y1 = int((y_center - height / 2) * scale_y)

        x2 = int((x_center + width / 2) * scale_x)
        y2 = int((y_center + height / 2) * scale_y)


        return {

            "class_name": "person",

            "confidence": best_confidence,

            "x1": x1,
            "y1": y1,

            "x2": x2,
            "y2": y2
        }



if __name__ == "__main__":


    project_directory = (
        Path(__file__).resolve().parent.parent
    )


    model_path = (
        project_directory
        / "models"
        / "yolov8n.onnx"
    )


    detector = YOLODetector(model_path)

File: cpp/src/detector.cpp
#include "detector.hpp"

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <iostream>
#include <stdexcept>
#include <vector>

#include <opencv2/imgproc.hpp>

#ifdef TP_HAVE_ONNX
#include <onnxruntime_cxx_api.h>
#endif

namespace tp {

#ifdef TP_HAVE_ONNX

/** Holds all ONNX Runtime state out of the public header. */
class YOLODetector::Impl {
public:
    Ort::Env env;
    Ort::SessionOptions options;
    Ort::Session session;

    std::vector<std::string> input_names;
    std::vector<std::string> output_names;
    std::vector<const char*> input_name_ptrs;
    std::vector<const char*> output_name_ptrs;

    int channel_count = 3;

    Impl()
        : env(ORT_LOGGING_LEVEL_WARNING, "tennis"),
          session(nullptr) {}

    // Ort::Session has no default constructor, so build it explicitly.
    void build(const std::string& path) {
        session = Ort::Session(env, path.c_str(), options);
    }
};

#else  // !TP_HAVE_ONNX

/** Empty placeholder so `sizeof(YOLODetector)` stays identical across TUs. */
class YOLODetector::Impl {};

#endif  // TP_HAVE_ONNX

YOLODetector::YOLODetector(const std::string& model_path,
                           int inference_size,
                           int threads)
    : model_path_(model_path),
      inference_size_(inference_size),
      threads_(threads) {
#ifdef TP_HAVE_ONNX
    impl_ = std::make_unique<Impl>();
    try {
        impl_->options.SetIntraOpNumThreads(std::max(1, threads_));
        impl_->options.SetGraphOptimizationLevel(
            GraphOptimizationLevel::ORT_ENABLE_ALL);
        impl_->build(model_path_);

        std::cout << "YOLO model loaded successfully" << std::endl;

        Ort::AllocatorWithDefaultOptions allocator;

        // Input information.
        std::vector<int64_t> input_shape;
        {
            auto type_info = impl_->session.GetInputTypeInfo(0);
            auto tensor_info = type_info.GetTensorTypeAndShapeInfo();
            input_shape = tensor_info.GetShape();
            impl_->input_names.emplace_back(
                impl_->session.GetInputNameAllocated(0, allocator).get());
        }

        // Use the model's own resolution when it declares one; the
        // constructor argument (Config.INFERENCE_SIZE) is the fallback.
        if (input_shape.size() >= 4 && input_shape[2] > 0) {
            inference_size_ = static_cast<int>(input_shape[2]);
        }

        // Output information.
        const size_t num_outputs = impl_->session.GetOutputCount();
        for (size_t i = 0; i < num_outputs; ++i) {
            impl_->output_names.emplace_back(
                impl_->session.GetOutputNameAllocated(i, allocator).get());
        }

        impl_->input_name_ptrs.reserve(impl_->input_names.size());
        for (const auto& name : impl_->input_names) {
            impl_->input_name_ptrs.push_back(name.c_str());
        }
        impl_->output_name_ptrs.reserve(impl_->output_names.size());
        for (const auto& name : impl_->output_names) {
            impl_->output_name_ptrs.push_back(name.c_str());
        }

        std::cout << "Input name: " << impl_->input_names[0] << std::endl;
        std::cout << "Input shape: [";
        for (size_t i = 0; i < input_shape.size(); ++i) {
            std::cout << (i == 0 ? "" : ", ") << input_shape[i];
        }
        std::cout << "]" << std::endl;
        std::cout << "Number of outputs: " << num_outputs << std::endl;
        for (size_t i = 0; i < num_outputs; ++i) {
            // Keep the Ort::TypeInfo alive while its (unowned) tensor shape
            // view is in use; otherwise the view dangles and dims read as
            // garbage. (GetTensorTypeAndShapeInfo returns an unowned view into
            // the TypeInfo, so the temporary in
            // `GetOutputTypeInfo(i).GetTensorTypeAndShapeInfo()` is a
            // use-after-free.)
            Ort::TypeInfo output_info = impl_->session.GetOutputTypeInfo(i);
            auto shape = output_info.GetTensorTypeAndShapeInfo().GetShape();
            std::cout << "Output: " << impl_->output_names[i] << " [";
            for (size_t j = 0; j < shape.size(); ++j) {
                std::cout << (j == 0 ? "" : ", ");
                if (shape[j] < 0) {
                    std::cout << "?";  // dynamic dimension
                } else {
                    std::cout << shape[j];
                }
            }
            std::cout << "]" << std::endl;
        }
    } catch (const std::exception& error) {
        throw std::runtime_error(
            "Failed to load ONNX model '" + model_path_ + "': " + error.what());
    }
#else
    (void)inference_size;
    (void)threads;
    throw std::runtime_error(
        "ONNX Runtime support not compiled in (TP_HAVE_ONNX). Rebuild with "
        "TP_ENABLE_ONNX=ON and a working onnxruntime install.");
#endif
}

YOLODetector::~YOLODetector() = default;

Detection YOLODetector::detectPerson(const cv::Mat& frame,
                                     float confidence_threshold) {
#ifdef TP_HAVE_ONNX
    if (!impl_ || !frame.data) {
        return Detection{};
    }

    const int size = inference_size_;
    const int plane = size * size;
    const int channels = impl_->channel_count;

    cv::Mat resized;
    cv::resize(frame, resized, cv::Size(size, size), 0, 0, cv::INTER_LINEAR);

    // BGR -> RGB, HWC -> CHW, normalized to [0, 1].
    std::vector<float> input(static_cast<std::size_t>(channels) * plane);
    const float inv = 1.0f / 255.0f;
    for (int y = 0; y < size; ++y) {
        for (int x = 0; x < size; ++x) {
            const cv::Vec3b& pixel = resized.at<cv::Vec3b>(y, x);
            const int offset = y * size + x;
            input[offset] = pixel[2] * inv;                    // R
            input[plane + offset] = pixel[1] * inv;            // G
            input[2 * plane + offset] = pixel[0] * inv;        // B
        }
    }

    const std::vector<int64_t> shape = {1, channels, size, size};
    Ort::MemoryInfo mem_info =
        Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
    Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
        mem_info, input.data(), input.size(), shape.data(), shape.size());

    auto outputs = impl_->session.Run(
        Ort::RunOptions(),
        impl_->input_name_ptrs.data(),
        &input_tensor,
        1,
        impl_->output_name_ptrs.data(),
        impl_->output_name_ptrs.size());

    if (outputs.empty()) {
        return Detection{};
    }

    Ort::Value& output = outputs[0];
    const auto out_info = output.GetTensorTypeAndShapeInfo();
    const auto out_shape = out_info.GetShape();
    const float* data = output.GetTensorData<float>();

    // Expect a YOLO-style head: [1, 84, N].
    if (out_shape.size() != 3) {
        return Detection{};
    }
    const int channels_out = static_cast<int>(out_shape[1]);
    const int num_anchors = static_cast<int>(out_shape[2]);
    if (channels_out <= 4 + PERSON_CLASS_ID) {
        return Detection{};
    }

    Detection best;
    float best_confidence = confidence_threshold;

    for (int n = 0; n < num_anchors; ++n) {
        auto at = [&](int k) -> float {
            return data[static_cast<std::size_t>(k) * num_anchors + n];
        };

        const float x_center = at(0);
        const float y_center = at(1);
        const float width = at(2);
        const float height = at(3);
        const float confidence = at(4 + PERSON_CLASS_ID);

        if (confidence < best_confidence) {
            continue;
        }
        best_confidence = confidence;

        const double scale_x = static_cast<double>(frame.cols) / size;
        const double scale_y = static_cast<double>(frame.rows) / size;

        best = Detection{};
        best.class_name = "person";
        best.confidence = confidence;
        best.x1 = static_cast<int>((x_center - width / 2) * scale_x);
        best.y1 = static_cast<int>((y_center - height / 2) * scale_y);
        best.x2 = static_cast<int>((x_center + width / 2) * scale_x);
        best.y2 = static_cast<int>((y_center + height / 2) * scale_y);
    }

    return best;
#else
    (void)frame;
    (void)confidence_threshold;
    return Detection{};
#endif
}

}  // namespace tp